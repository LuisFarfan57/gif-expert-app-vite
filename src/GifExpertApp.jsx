import { useState } from 'react'
import { AddValue, GifGrid } from './components'

const GifExpertApp = () => {
  const [categories, setCategories] = useState([])

  const addCategory = (category) => {
    if (categories.map(category => category.toLowerCase()).includes(category.toLowerCase())) return
    setCategories([category, ...categories])
  }

  return (
    <>
      <h1>GifExpertApp</h1>

      {/* usar la palabra on cuando se esta emitiendo un evento */}
      <AddValue onNewValue={addCategory} placeholder="Write the new category" />

      {categories.map((category) => (
        <GifGrid key={category} category={category} />
      ))}
    </>
  )
}

export default GifExpertApp
