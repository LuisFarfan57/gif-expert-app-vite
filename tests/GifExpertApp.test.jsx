import { fireEvent, render, screen } from "@testing-library/react"
import GifExpertApp from "../src/GifExpertApp"

describe('GifExpertApp tests', () => {
  it('renders correctly', () => {
    const { container } = render(<GifExpertApp />)

    expect(container).toMatchSnapshot()
  })

  const category1 = '  One Punch Man  '
  const category2 = '  Boku no hero  '
  const categoryRepeated = 'BoKu NO HERO'

  it('adds new categories in the h3 headings when we enter new categories', () => {
    render(<GifExpertApp />)

    const input = screen.getByRole('textbox')
    fireEvent.input(input, { target: { value: category1 } })
    fireEvent.keyUp(input, { key: 'Enter' })
    fireEvent.input(input, { target: { value: category2 } })
    fireEvent.keyUp(input, { key: 'Enter' })

    // And also is trimmed
    const headings = screen.getAllByRole('heading', { level: 3 })
    expect(headings.length).toBe(2)
    expect(headings[0].innerHTML).toBe(category2.trim())
  })

  it('do not adds the new category if it is repeated', () => {
    render(<GifExpertApp />)

    const input = screen.getByRole('textbox')
    fireEvent.input(input, { target: { value: category1 } })
    fireEvent.keyUp(input, { key: 'Enter' })
    fireEvent.input(input, { target: { value: category2 } })
    fireEvent.keyUp(input, { key: 'Enter' })
    fireEvent.input(input, { target: { value: categoryRepeated } })
    fireEvent.keyUp(input, { key: 'Enter' })

    // And checks that even if it is in different case, doesn't add the category
    expect(screen.getAllByRole('heading', { level: 3 }).length).toBe(2)
  })
})
