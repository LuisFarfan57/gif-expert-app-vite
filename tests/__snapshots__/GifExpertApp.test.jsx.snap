// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifExpertApp tests renders correctly 1`] = `
<div>
  <h1>
    GifExpertApp
  </h1>
  <input
    name="newValueInput"
    placeholder="Write the new category"
    type="text"
    value=""
  />
</div>
`;
